tests/ops/op_math/test_sprod.py

```diff
@@ -0,0 +1,635 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for the SProd class representing the product of an operator by a scalar"""
+
+import pytest
+import numpy as np
+from copy import copy
+
+from scipy.sparse import csr_matrix
+
+import pennylane as qml
+from pennylane import math
+import pennylane.numpy as qnp
+
+from pennylane.wires import Wires
+from pennylane import QuantumFunctionError
+from pennylane.ops.op_math.sprod import s_prod, SProd
+from pennylane.ops.op_math.sprod import _sprod  # pylint: disable=protected-access
+from pennylane.operation import MatrixUndefinedError, DecompositionUndefinedError
+import gate_data as gd  # a file containing matrix rep of each gate
+
+scalars = (1, 1.23, 0.0, 1 + 2j)  # int, float, zero, and complex cases accounted for
+
+no_mat_ops = (
+    qml.Barrier,
+    qml.WireCut,
+)
+
+non_param_ops = (
+    (qml.Identity, gd.I),
+    (qml.Hadamard, gd.H),
+    (qml.PauliX, gd.X),
+    (qml.PauliY, gd.Y),
+    (qml.PauliZ, gd.Z),
+    (qml.S, gd.S),
+    (qml.T, gd.T),
+    (qml.SX, gd.SX),
+    (qml.CNOT, gd.CNOT),
+    (qml.CZ, gd.CZ),
+    (qml.CY, gd.CY),
+    (qml.SWAP, gd.SWAP),
+    (qml.ISWAP, gd.ISWAP),
+    (qml.SISWAP, gd.SISWAP),
+    (qml.CSWAP, gd.CSWAP),
+    (qml.Toffoli, gd.Toffoli),
+)
+
+param_ops = (
+    (qml.RX, gd.Rotx),
+    (qml.RY, gd.Roty),
+    (qml.RZ, gd.Rotz),
+    (qml.PhaseShift, gd.Rphi),
+    (qml.Rot, gd.Rot3),
+    (qml.U1, gd.U1),
+    (qml.U2, gd.U2),
+    (qml.U3, gd.U3),
+    (qml.CRX, gd.CRotx),
+    (qml.CRY, gd.CRoty),
+    (qml.CRZ, gd.CRotz),
+    (qml.CRot, gd.CRot3),
+    (qml.IsingXX, gd.IsingXX),
+    (qml.IsingYY, gd.IsingYY),
+    (qml.IsingZZ, gd.IsingZZ),
+)
+
+ops = (
+    (1.0, qml.PauliX(wires=0)),
+    (0.0, qml.PauliZ(wires=0)),
+    (1j, qml.Hadamard(wires=0)),
+    (1.23, qml.CNOT(wires=[0, 1])),
+    (4.56, qml.RX(1.23, wires=1)),
+    (1.0 + 2.0j, qml.Identity(wires=0)),
+    (10, qml.IsingXX(4.56, wires=[2, 3])),
+    (0j, qml.Toffoli(wires=[1, 2, 3])),
+    (42, qml.Rot(0.34, 1.0, 0, wires=0)),
+)
+
+ops_rep = (
+    "1.0*(PauliX(wires=[0]))",
+    "0.0*(PauliZ(wires=[0]))",
+    "1j*(Hadamard(wires=[0]))",
+    "1.23*(CNOT(wires=[0, 1]))",
+    "4.56*(RX(1.23, wires=[1]))",
+    "(1+2j)*(Identity(wires=[0]))",
+    "10*(IsingXX(4.56, wires=[2, 3]))",
+    "0j*(Toffoli(wires=[1, 2, 3]))",
+    "42*(Rot(0.34, 1.0, 0, wires=[0]))",
+)
+
+
+def get_qft_mat(num_wires):
+    """Helper function to generate the matrix of a qft protocol."""
+    omega = math.exp(np.pi * 1.0j / 2 ** (num_wires - 1))
+    mat = math.zeros((2**num_wires, 2**num_wires), dtype="complex128")
+
+    for m in range(2**num_wires):
+        for n in range(2**num_wires):
+            mat[m, n] = omega ** (m * n)
+
+    return 1 / math.sqrt(2**num_wires) * mat
```

Comment:
QFT does define its matrix, and I think it's fine to rely on it here. Just use `QFT.compute_matrix` instead.

Reply:
got it !